### DISCUSSION TITLE
Alternate language when writing in Quarto book format

### DISCUSSION THREAD

**Discussion by KennethGottfredsen** [quarto-dev/quarto-cli#2453] [Q&A]
I'm trying to code in the Quarto book format. 

How do I override the values for the “Author” and “Published” captions when writing a book project?

`---
title: "Min bog"
author: "Kenneth Gottfredsen"
date: 5/22/2022
language: 
  title-block-author-single: "Forfatter"
  title-block-published: "Opdateret"
---`

Above works when I'm typing in the document format, but it doesn't work when I copy and paste it into the yaml of a book project.

**Comment by cderv** [comment]
> Above works when I'm typing in the document format, but it doesn't work when I copy and paste it into the yaml of a book project.

Where did you put the language field in the `_quarto.yml` file ? It should be placed at top level I believe. 

Also you could have a look at new feature for our current pre-release, about project profile: https://quarto.org/docs/projects/profiles.html#language-content

**Reply by KennethGottfredsen** [reply]
Greetings cderv,

I don't have any language field specified, as far as I know Quarto don't support the Danish language anyway. It would be nice to reproduce my papers in my native language instead of English.  

Basically what I want to do is the following:

1. Remove or at least rename the default "author" to forfatter and "published" to udgivet. 
2. Rename the table of contents to "Indholdsfortegnelse". 

This is my code copied directly from the yaml document:

project:
  type: book

book:
  title: "Studiebog"
  author: "Kenneth Gottfredsen"
  #author-title: "Kenneth Gottfredsen"
  date: "18.9.2022"
  chapters:
    - index.qmd
    - intro.qmd
    - summary.qmd
    - references.qmd

bibliography: references.bib

format:
  html: 
    theme: darkly
  pdf:
    documentclass: scrreprt

editor: visual


I hope you know what I mean. Feel free to ask more questions. 

/ Kenneth

**Reply by cderv** [reply]
See the doc about language then: https://quarto.org/docs/authoring/language.html

* You can use alternate language for some fields using YAML value
* Or you can even provide your own custom translation by providing a full YAML file. 

Everything is in the doc - look at current translation file for know field values supporting translation. Is that not working ?

And please do contribute Danish language in Quarto for native support of your language! https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/language#contributing-localizations

Thank you !

**Comment by KennethGottfredsen** [comment]
Hi again cderv, 

I forgot to mention one more request. 

How do I adjust the code chunks to include install.package('tidyverse') or pload::packman, to include the "copy to clipboard" option? When I type the function in an R code chunk I get an error. I need to do an assignment, and I want to show that I'm installing packages loading from the library. I want my readers to just be able to copy and paste my chunks directly. Is it because of my format is html?  

/ Kenneth

**Reply by cderv** [reply]
please open a new thread for new issue. You can then ask your question by sharing some examples. 

Some quick hints though: 
* Inserting chunk with `#| eval: FALSE` chunk option will show the chunk without evaluating (which you usually want to do for explaining something like installating package, and not installing them during render)
* [code copy button](https://quarto.org/docs/output-formats/html-code.html#copy-button) should be the default in HTML format.

**Comment by fredguth** [comment]
I found this question because I am also trying to change language of a book.

I followed instructions from documentation, but it didn't work:
```_quarto.yml
project:
  type: book

book:
  title: "My book title"
  language: "pt-BR"
  chapters:
    - index.qmd
format:
  html:
    theme: cosmo
````

**Reply by fredguth** [reply]
I also tried using `lang: pt-BR` and `language: fr` just to test--- nothing.

**Reply by cderv** [reply]
It seems there is indeed an issue as `language` is among the keys suggested under `book`. 

It should be with `lang` and specified at first level in your YAML. Please try this. 
```` 
lang: pt-BR
````

`language` is for tweaking a default translation, or providing a custom one using a language yaml file (Doc: https://quarto.org/docs/authoring/language.html)

**Reply by cderv** [reply]
I opened https://github.com/quarto-dev/quarto-cli/issues/2707 for looking into this

**Reply by fuhrmanator** [reply]
I'm also having a problem with non-breaking spaces before punctuation (e.g., before `:`) in the PDF in French (this normally works great with babel, and the quarto documentation says the `lang` option is supposed to trigger babel translations). I tried forcing `\usepackage[french]{babel}` in the headers, but it seems it's already being set (there's a LaTeX conflict). 

I can't debug the .latex (anymore?) because when I try to render my book project `--to latex`, it says it's not supported. How can I debug the .tex files produced before the PDF of a book?

ETA: I remember now what's happening with the spacing. Pandoc's default template turns off the feature (or at least it used to): https://tex.stackexchange.com/a/505068/17868

ETA2: I got it to work, by overriding the [Template](https://quarto.org/docs/journals/templates.html#templates) for LaTeX. I actually took all the partials into my project and told Quarto to use the top-level `template.tex`:

```yaml
format:
  pdf:
    template: partials/template.tex
```

It's somewhat tedious to copy all those partials, so I just downloaded a .zip of the quarto-cli code and copied the folder.

I had to make the following changes in the `template.tex` file to get it to work:

```tex
$if(lang)$
\ifLuaTeX
\usepackage[bidi=basic,french]{babel}
\else
\usepackage[bidi=default,french]{babel}
\fi
$if(babel-lang)$
\babelprovide[main,import]{$babel-lang$}
$endif$
$for(babel-otherlangs)$
\babelprovide[import]{$babel-otherlangs$}
$endfor$
% get rid of language-specific shorthands (see #6817):
% \let\LanguageShortHands\languageshorthands
% \def\languageshorthands#1{}
$endif$
```

Here's an annotation of the changes:

![image](https://user-images.githubusercontent.com/7606540/196853151-06a148c8-8e35-49fd-9265-3c9f52704d01.png)

**Reply by cderv** [reply]
@fuhrmanator thanks for sharing. I believe you are finding some limitation or misbehavior from Pandoc itself. The change you are making is directly related to some Pandoc's default template content. 

Looking into this, I am finding that the line you commented should now be removed by Pandoc : https://github.com/jgm/pandoc/issues/6817 is still open for this purpose (https://github.com/jgm/pandoc/issues/6817#issuecomment-850891209) - you should comment there maybe. 

Regarding the initial issue of French rule, this is also an open Pandoc issue you found - see https://github.com/jgm/pandoc/issues/8283 for discussion. You should also contribute there. 

When this will be fixed, then we can update template in Quarto. 

A note from the last issue mentioned: There is a Lua filter you can probably used for non breaking space
https://github.com/InseeFrLab/pandoc-filter-fr-nbsp made by @RLesur. @RLesur, how to make this Lua filter available as an extension to use more easily with Quarto ? 

Thank you

**Reply by fuhrmanator** [reply]
Thanks for tracking down those details. After some reading, it seems to me that this will take a long time to get fixed, because it's not just pandoc, but also the problematic French support in babel (.ldf vs .ini). It's a great example of complexity due to internationalization. 

I like the idea of creating a Lua filter like @RLesur has done, since other formats would take advantage of it. However, it's important to point out a difference: the philosophy in [frenchb](https://ctan.mirror.globo.tech/macros/latex/contrib/babel-contrib/french/frenchb.pdf) (LaTeX bable French) is not to use any spaces before punctuation in the source. It's extra work to insert them in the LaTeX. So, you can simply type `Bonjour!` in your LaTeX and frenchb will insert the non-breaking space before the punctuation (and after it, in some cases e.g. `:`). It's pretty detailed support according to French punctuation rules. There's even support for non-breaking spaces after `«` and before `»` (which I have managed to get already by hacking the https://github.com/odkr/pandoc-quotes.lua/blob/main/pandoc-quotes.lua). 

Looking at the tests for https://github.com/InseeFrLab/pandoc-filter-fr-nbsp, I think this filter is much more rudimentary. It expects you've already written `Bonjour !` and it converts the space before certain punctuation to non-breaking. I'm not enough of a Lua/Pandoc expert to know how hard it would be to make it work like babel/French, but it's easy to write up some tests with a frenchb-like support.

**Reply by fuhrmanator** [reply]
I hope you don't mind me posting them in this thread, but I found another "quirk" relating to `lang:` setting in dates that have names of months (not just for books). I'm on windows, so maybe it's OS specific? Here's a MWE:

```markdown
---
title: "MWE"
lang: fr-FR
date: last-modified
date-format: "DD MMMM YYYY à HH:mm:ss Z"
---

Here's some text.
```

Rendering this in HTML or PDF will show a date that is not French (note "October" rather than "octobre"). 

> MWE
> DATE DE PUBLICATION
> 26 October 2022 à 09:59:56 -04:00
>
> Here’s some text.


The workaround is to drop the country from the language, e.g., `lang: fr` -- it's probably another bug with pandoc?

> MWE
> DATE DE PUBLICATION
> 26 octobre 2022 à 10:00:57 -04:00
> 
> Here’s some text.

**Reply by cderv** [reply]
@fuhrmanator I don't know exactly about that one. I would open an issue with this small reprex. I believe @dragonstyle will then be able to look into this. Seems like we don't handle that all correctly to me. Thank you !

**Reply by dragonstyle** [reply]
Yes this is definitely a bug with the way we load date formatting locales- we have locales for `fr`, `fr-ca`, and `fr-ch`, but we are missing fallback logic for the case when we don't have a region specific locale that matches, but we do have a general language matching locale...

**Reply by dragonstyle** [reply]
I've created an issue here:

https://github.com/quarto-dev/quarto-cli/issues/3059

**Reply by cderv** [reply]
Awesome thanks you @dragonstyle and thank you @fuhrmanator for the report!

**Comment by fredguth** [comment]
For any one trying to solve this, create a `_metadata.yml` in your root directory with:
```
lang: pt-BR
```
As the `language` option in the book project is just for the csl (citations)  (see https://github.com/quarto-dev/quarto-cli/issues/2707)

**Reply by fuhrmanator** [reply]
`language:` works for me on my book project to customize the translations https://quarto.org/docs/authoring/language.html#custom-translations

**Reply by fredguth** [reply]
Could you share your repo? If not I will try to reproduce a Minimum Working Example of the problem I was getting.

**Reply by fuhrmanator** [reply]
@fredguth my current repo doesn't use it anymore because I am trying to minimize the use of customizations. But at this point https://github.com/fuhrmanator/log210-ndc-quarto/blob/1664439764e76a173e8403d3c6dcff695fae6dec/_quarto.yml it was working as far as I know.

**Reply by fuhrmanator** [reply]
This is the line number: https://github.com/fuhrmanator/log210-ndc-quarto/blob/1664439764e76a173e8403d3c6dcff695fae6dec/_quarto.yml#L11
